### PR TITLE
KAFKA-12339: Add retry to admin client's listOffsets

### PR DESCRIPTION
**Pull request by chia7712**
issue: https://issues.apache.org/jira/browse/KAFKA-12339

After upgrading our connector env to 2.9.0-SNAPSHOT, sometimes the connect cluster encounters following error.
```
Uncaught exception in herder work thread, exiting:  (org.apache.kafka.connect.runtime.distributed.DistributedHerder:324)
org.apache.kafka.connect.errors.ConnectException: Error while getting end offsets for topic 'connect-storage-topic-connect-cluster-1'
at org.apache.kafka.connect.util.TopicAdmin.endOffsets(TopicAdmin.java:689)
at org.apache.kafka.connect.util.KafkaBasedLog.readToLogEnd(KafkaBasedLog.java:338)
at org.apache.kafka.connect.util.KafkaBasedLog.start(KafkaBasedLog.java:195)
at org.apache.kafka.connect.storage.KafkaStatusBackingStore.start(KafkaStatusBackingStore.java:216)
at org.apache.kafka.connect.runtime.AbstractHerder.startServices(AbstractHerder.java:129)
at org.apache.kafka.connect.runtime.distributed.DistributedHerder.run(DistributedHerder.java:310)
at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.UnknownTopicOrPartitionException: This server does not host this topic-partition.
at org.apache.kafka.common.internals.KafkaFutureImpl.wrapAndThrow(KafkaFutureImpl.java:45)
at org.apache.kafka.common.internals.KafkaFutureImpl.access$000(KafkaFutureImpl.java:32)
at org.apache.kafka.common.internals.KafkaFutureImpl$SingleWaiter.await(KafkaFutureImpl.java:89)
at org.apache.kafka.common.internals.KafkaFutureImpl.get(KafkaFutureImpl.java:260)
at org.apache.kafka.connect.util.TopicAdmin.endOffsets(TopicAdmin.java:668)
... 10 more
```
#9780 added shared admin to get end offsets. KafkaAdmin#listOffsets does not handle topic-level error, hence the UnknownTopicOrPartitionException on topic-level can obstruct worker from running when the new internal topic is NOT synced to all brokers.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by rhauch** [comment]
Thanks, @chia7712. I've reviewed much of this, and it seems straightforward, though I plan to review more thoroughly tomorrow (~12 hours). I've asked a few others that might be more familiar with the admin client code to also take a look.

**Review by rhauch — APPROVED** [review]
LGTM. Thanks, @chia7712.

File: clients/src/main/java/org/apache/kafka/clients/admin/internals/MetadataOperationContext.java
Comment (kkonstantine) [inline clients/src/main/java/org/apache/kafka/clients/admin/internals/MetadataOperationContext.java:85]: nit: a stylistic observation is that this call is exactly the same as the call below. Yet it's written differently. 

Styles differ, usually not too much, from module to module. I think it's good to keep existing styles to help with readability when the changes don't require a greater change. 
 
```suggestion
            if (shouldRefreshMetadata(tm.error())) {
                throw tm.error().exception();
            }
```

**Review by kkonstantine — APPROVED** [review]
The changes LGTM. 
Left a minor comment. 

Also might be good to note in the description that the error could be retriable in a same way as the one on partition metadata and it's the fact that it is not retried (here or in the connect worker) that creates the issue.

**Comment by chia7712** [comment]
@rhauch @kkonstantine Thanks for your reviews!

**Comment by chia7712** [comment]
> Also might be good to note in the description that the error could be retriable in a same way as the one on partition metadata and it's the fact that it is not retried (here or in the connect worker) that creates the issue.

Should we cherry-pick this patch to other active branches?

**Comment by rhauch** [comment]
I've already cherry picked this to `2.7`, `2.6` (after release manager's approval) and `2.5`, and have asked for approval on `2.8` (since we are in code freeze).

**Pull request merged**
